### PR TITLE
Finish Firestore Watch implementation

### PR DESCRIPTION
**Pull request by jskeet**
It's probably worth reviewing the first commit, and we could potentially merge that while we discuss everything else - or keep it in this PR.

Initially I'm looking for feedback on the structure:

- Does the public API look correct?
- Does the separation of SnapshotListener from WatchStream make sense, or should I combine them?
- Does the shape of WatchStream.Start make sense, even if some async aspects may be wrong?

(Any implementation aspects which are clearly broken would be good to know too...)

### CODE REVIEW

**Comment by jskeet** [comment]
(Hold off reviewing for now - I have a little sample, but it's not working yet...)

**Comment by jskeet** [comment]
Turns out it was a typo in my test. I've added the test in a new commit. Sample output:

```text
Creating document for Sophie (Score = 7)
Sophie document ID: QNBqD8B0T0AAc7tSi6q5
Callback received snapshot
Count: 1
Changes:
QNBqD8B0T0AAc7tSi6q5: ChangeType=Added; OldIndex=; NewIndex=0)

Creating document for James (Score = 10)
James document ID: Vjspdol0s0fVoTZToM95
Callback received snapshot
Count: 2
Changes:
Vjspdol0s0fVoTZToM95: ChangeType=Added; OldIndex=; NewIndex=0)

Modifying document for Sophie (set Score = 11, higher than score for James)
Callback received snapshot
Count: 2
Changes:
QNBqD8B0T0AAc7tSi6q5: ChangeType=Modified; OldIndex=1; NewIndex=0)

Modifying document for James (set Score = 4, below threshold for query)
Callback received snapshot
Count: 1
Changes:
Vjspdol0s0fVoTZToM95: ChangeType=Removed; OldIndex=1; NewIndex=)

Deleting document for Sophie
Callback received snapshot
Count: 0
Changes:
QNBqD8B0T0AAc7tSi6q5: ChangeType=Removed; OldIndex=0; NewIndex=)

Stopping listener
```

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/SnapshotListener.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    /// <summary>
+    /// An ongoing "listen" or "watch" operation on either document or query snapshots.
+    /// This is returned from <c>Listen</c> methods.
+    /// </summary>
+    public sealed class SnapshotListener
+    {
+        private readonly WatchStream _stream;
+        private int _stopped;
+
+        // TODO: Rename! (Don't know what to though...)
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/SnapshotListener.cs]: ListenerTask?
WatchTask?
Agreed that `Task` isn't the nest name ;)

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/DocumentReference.cs [hunk 2]
```diff
@@ -208,5 +208,34 @@ public Task<WriteResult> UpdateAsync(string field, object value, Precondition pr
         /// </summary>
         /// <returns>A lazily-iterated sequence of collection references within this document.</returns>
         public IAsyncEnumerable<CollectionReference> ListCollectionsAsync() => Database.ListCollectionsAsync(this);
+
+        // TODO: Should this be ListenAsync?
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/DocumentReference.cs]: Don't think so, this method doesn't itself block.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/WatchStream.cs [hunk 3]
```diff
@@ -12,11 +12,235 @@
...
+        // cancellation token provided to Listen, or the cancellation token provided to StopAsync.
+        private readonly CancellationTokenSource _callbackCancellationTokenSource;
+        private readonly IWatchState _state;
+        private readonly RetrySettings.IJitter _backoffJitter;
+        private readonly BackoffSettings _backoffSettings;
+        private readonly Target _target;
+        private readonly object _stateLock = new object();
+
+        // Only mutable variable within the class: if the user calls Stop and passes in a cancellation token,
+        // we link it with _callbackCancellationTokenSource, and need to be able to dispose of that regist
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/WatchStream.cs]: nit: Missing end of word/sentence!

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/WatchStream.cs [hunk 4]
```diff
@@ -12,11 +12,235 @@
...
+                        // Otherwise, close the current stream and restart.
+                        await CloseStreamAsync().ConfigureAwait(false);
+                        cause = StreamInitializationCause.StreamCompleted;
+                    }
+                    catch (RpcException e) when (s_transientErrorStatusCodes.Contains(e.Status.StatusCode))
+                    {
+                        // Close the current stream, ready to create a new one.
+                        await CloseStreamAsync().ConfigureAwait(false);
+                        // Extend the back-off if necessary.
+                        if (e.Status.StatusCode == StatusCode.ResourceExhausted)
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/WatchStream.cs]: I looks dangerous to not increase the backoff at all on most conditions.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/WatchStream.cs [hunk 5]
```diff
@@ -12,11 +12,235 @@
...
+                    }
+                }
+            }
+
+            async Task CloseStreamAsync()
+            {
+                if (underlyingStream != null)
+                {
+                    var completeTask = underlyingStream.TryWriteCompleteAsync();
+                    // TODO: Handle exceptions from this?
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/WatchStream.cs]: Agreed, not sure what's best in this situation.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore/DocumentReference.cs [hunk 6]
```diff
@@ -208,5 +208,34 @@ public Task<WriteResult> UpdateAsync(string field, object value, Precondition pr
...
+        /// <summary>
+        /// Watch this document for changes.
+        /// </summary>
+        /// <param name="callback">The callback to invoke each time the document changes. Must not be null.</param>
+        /// <param name="cancellationToken">Optional cancellation token which may be used to cancel the listening operation.</param>
+        /// <returns>A <see cref="SnapshotListener"/> which may be used to monitor the listening operation and stop it gracefully.</returns>
+        public SnapshotListener Listen(Func<DocumentSnapshot, CancellationToken, Task> callback, CancellationToken cancellationToken = default)
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore/DocumentReference.cs]: I'm curious - why are these methods called `Listen` and not `Watch`?

Comment (jskeet) [reply apis/Google.Cloud.Firestore/Google.Cloud.Firestore/DocumentReference.cs]: Watch is the internal name, but Listen is the name determined for the public API. Personally I like Watch, but it's better to be consistent with all the other languages :)

**Comment by jskeet** [comment]
I've added a bit of testing - I'll create more tests if you're happy with the structure. There was a lot of faking to do, and it's annoying that I couldn't use Moq in the IWatchState, but it all does appear to work.

Currently I'm not testing the timing at all - that could potentially be done later.

**Comment by jskeet** [comment]
Okay, I think this is now ready for review with an eye to actually merging.

File: apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/WatchStreamTest.cs [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+                var call = (Action<StreamInitializationCause>) _expectedCalls.Dequeue();
+                call(cause);
+            }
+
+            internal void AddResumeToken(int index, ByteString token) => _resumeTokens.Add((index, token));
+
+            internal void ExpectResponse(ListenResponse expectedResponse, WatchResponseResult result)
+            {
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+                Func<ListenResponse, CancellationToken, Task<WatchResponseResult>> call = async (response, _) =>
```
Comment (chrisdunelm) [inline apis/Google.Cloud.Firestore/Google.Cloud.Firestore.Tests/WatchStreamTest.cs]: Why not write this as a non-async lambda that returns `Task.FromResult(0)` ?

**Pull request merged**